### ISSUE TITLE
Heterogenous Workloads

### ISSUE THREAD

**Issue by kflansburg** [deislabs/krustlet#297]
Support for specifying pods with both lxc and wasm workloads, which then execute in a single namespace. This could allow running sidecar containers with wasm workloads (such as Istio). 

This probably involves a couple of features:

- [ ] Support for multiple providers in a single Kubelet (for example OCI + WASM). 
- [ ] Spawn WASM workloads in a specific namespace (the one use by the container workloads). 
- [ ] Dispatch pod spec to multiple providers and merge status from each back to ApiServer.
- [ ] Dispatch API calls like `logs` and `exec`.
- [ ] Coordinate cleaning up heterogenous namespace on Pod delete.
- [ ] I'm not sure what options exist for Windows support.

**Comment by bacongobbler** [comment]
> Support for multiple providers in a single Kubelet (for example OCI + WASM).

In my view I see this as a separate provider that can handle both types of workload rather than two separate providers. Reason being is that the architecture from the project's perspective is much simpler: we only have to be concerned about the Provider API, and one has to consume the API in their own fashion to handle these use cases. There may also be some assembly required for a WebAssembly module and a container to share the same network/filesystem namespace, and some features available for OCI that may not have an equivalent for WebAssembly (cgroups, for example).

The trickiest part will be to determine the image manifest type before pulling the image. That way we can determine whether the pod is requesting a container or a WebAssembly module. I don't think we have that mechanism available in `oci-distribution` right now.

>     * I'm not sure what options exist for Windows support.

There are implementations of kubelets running on a Windows node running Windows Containers. I'm assuming the same could be applied here. Once Window support is complete in #20, a provider with Windows Container and WebAssembly support could be neat.

https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/adding-windows-nodes/#getting-started-adding-a-windows-node-to-your-cluster

**Comment by kflansburg** [comment]
I agree that implementing this logic in Kubelet doesn’t make much sense. Perhaps this could be a Provider instead which manages other homogeneous providers so that we do not have to reimplement the actual execution of WASM and LXC workloads.

**Comment by bacongobbler** [comment]
A provider's provider. I like it. 👍

**Comment by itowlson** [comment]
I started looking at this because it seems like it would be really useful - the service mesh example had been on my mind too.

One difficulty I immediately ran up against was what to advertise as the `kubernetes.io/arch`.  If a kubelet can run both containerised x86 binaries _and_ WASM modules - as in the example of a WASM container with an Istio sidecar - then what architecture should it claim to be?  Is there a way to advertise multiple architectures?  (I tried to figure our what Virtual Kubelet did for this but couldn't decipher it.)

**Comment by kflansburg** [comment]
I cannot recall where, but I believe I read somewhere that the Kubernetes spec for a Pod explicitly disallows something like this. So we are fighting the API a bit on this.

I haven’t used Virtual Kublet, does it allow you to schedule containers on multiple types of compute in a single pod?

If we go forward with the meta provider idea, our architecture could simply be a tuple that matches a given meta provider and the architectures that it supports in some predefined ordering.

**Comment by itowlson** [comment]
The `arch` label from our point of view is at the Node level, not the Pod level - the virtual node implemented by the kubelet currently advertises an architecture.  I'm not sure if a node is allowed to advertise multiple architectures...

**Comment by kflansburg** [comment]
Well the pod makes requests for a given architecture. So they must both agree. It may have been Nodes that I read the limitation about though.

**Comment by itowlson** [comment]
@bacongobbler I think the composite provider-of-providers might be tricky with our current design.  Adding a heterogeneous pod to the composite provider would require it to dispatch _individual containers_ to the architecture-specific providers, which isn't currently possible (providers operate only on pods).  So we might need a lower-level 'container provider' abstraction; existing providers such as WASI would host only one container provider, but a composite provider could host several.

(Although it might be desirable to always support traditional container workloads even if the kubelet says it doesn't, because of the `kube-system` pods that get scheduled regardless of taints.)

**Comment by kflansburg** [comment]
I think my plan was to modify the pod definition to create two pod definitions containing only containers from each architecture and then forward them to the respective providers. I do think things will get messy (especially when the providers communicate back to the api server), but was planning to give that a shot in the next few days.
